lib/hocon/cli.rb

```diff
@@ -188,7 +193,11 @@ def self.invalid_subcommand_error(subcommand, opt_parser)
   # If out_file is not nil, write to that file. Otherwise print to STDOUT
   def self.print_or_write(string, out_file)
     if out_file
-      File.open(out_file, 'w') { |file| file.write(string) }
+      File.open(out_file, 'w') do |f|
+        f.flock(File::LOCK_EX)
+        f.write(string)
+        f.flock(File::LOCK_UN)
+      end
```

Comment:
I don't think that doing the locking this way will actually prevent the race conditions that @binford2k was getting at.

Say that you have two process attempt to modify the file at exactly the same time.  What would happen with this code is that both processes could get the shared lock and read the original file into memory, and then one gets the write lock, writes out its changes, and then the second one gets the write lock and writes out its changes, which will override anything done by the first process.  So it's a "last wins" scenario.

The locking that you have here would theoretically prevent two simultaneous attempts to write from corrupting the file, which is nice.

I'm also not sure what happens if an exception occurs while the lock is held.  Is it possible that the process leaves the file locked and then subsequent calls to the tool just hang?

Overall I think that locking is a complex enough topic that we should probably just leave it out for now, and file a second ticket to adding it in later.  @binford2k have you run into real-world situations where races between multiple processes trying to edit a HOCON file have actually caused problems?